Update Aggregator and AggregatorFactory interfaces to improve mem estimates

kfaraz opened the pull request:
Fixes #12022  

### Description

The current implementation of memory estimation in `OnHeapIncrementalIndex`:
- uses `guessAggregatorHeapFootprint()` to calculate the max memory an metric aggregator can use
- multiplies the above value by the number of aggregators (same as number of aggregated rows or number of unique row keys)

Current implementation of `StringDimensionIndexer`
- encodes dimension values and calculates size of encoded values as well as original dimension values every time, irrespective of whether the dimension values are already added to the dictionary

Because of the above, the memory usage tends to be over-estimated which leads to more persistence cycles than necessary.

This PR replaces the max estimation mechanism with getting the actual incremental memory used by the aggregator or indexer at each invocation of `aggregate` or `encode` respectively.

<hr>

### Changes
- Add new flag `useMaxMemoryEstimates` in the task context. This overrides the same flag in DefaultTaskConfig i.e. `druid.indexer.task.default.context` map
- `useMaxMemoryEstimates = true`(default value) denotes the current method of estimation
- Add method `AggregatorFactory.factorizeWithSize()` that returns an `AggregatorAndSize` which contains
  - the `Aggregator` instance
  - initial size in bytes of the aggregator
- Add method `Aggregator.aggregateWithSize()`
  - returns a long representing the incremental memory used by this aggregation
  - default implementation calls `aggregate()` and returns 0
- Remove method `DimensionIndexer.estimateEncodedKeyComponentSize()`
- Update the method `DimensionIndexer.processRowValsToKeyComponent()` to return `EncodedKeyComponent<EncodedType>` which contains:
    - `EncodedType keyComponent`: e.g. int[] for StringDimensionIndexer, Long for LongDimensionIndexer
    - `long effectiveSizeBytes`: Effective size of the key component.
- Update `OnHeapIncrementalIndex` to use the new estimations only if `useMaxMemoryEstimates = false`
- Update `Aggregator` impls

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregator.java:
```diff
@@ -64,6 +86,25 @@ public void aggregate()
   }
 
+  @Override
+  public long aggregateWithSize()
```
As I read this code, there's 2 different "lifecycle"s happening.  One where normal `aggregate()` is being called and using `initUnion()` and another where `aggregateWithSize()` is being called and expecting something external to have called `getInitialSizeBytes()`.

I find myself wondering if it wouldn't be nicer to have two concrete classes each specialized to each lifecycle and not trying to mix the code together.  There's probably still a lot of room for re-use, but with the current mixing of the code, it seems relatively simple to accidentally use one when the other should've been used (i.e. if we separate out the classes, then the `aggregate()` method for the sized one can throw an `UnsupportedOperationException` which will ensure that it's never called when sizing is supposed to be used.

kfaraz replied:
Re-structured to remove dependence between the methods `initUnion()` and `aggregateWithSize()`

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregator.java:
```diff
@@ -133,4 +174,34 @@ static void updateUnion(Union union, Object update)
+
+    // Initialize the sketch if not already initialized
+    if (union == null) {
+      initUnion();
```
This feels a little bit wrong to me to init on the initial sizing.  If it starts out `null`, then its size is `0`.  It seems like we should either do this lazily, OR we should initialize the union on construction.

kfaraz replied:
Re-structured the code to initialize lazily and account for the size only when initializing.

imply-cheddar commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -849,6 +849,8 @@ private TaskStatus generateAndPublishSegments(
     final PartitionsSpec partitionsSpec = partitionAnalysis.getPartitionsSpec();
     final IndexTuningConfig tuningConfig = ingestionSchema.getTuningConfig();
     final long pushTimeout = tuningConfig.getPushTimeout();
+    tuningConfig.useMaxMemoryEstimates = getContextValue(
```
If it's on the tuning config, then it can be taken as another parameter from the tuning config's JSON and doesn't need to be on the context.  Setting values on the config like this is a bit dangerous, so let's just take it on the constructor and stop looking at the context of the Task.

imply-cheddar commented on processing/src/main/java/org/apache/druid/query/aggregation/SizedAggregator.java:
```diff
@@ -0,0 +1,49 @@
+ * Encapsulates an {@link Aggregator} and the initial size in bytes required by
+ * the Aggregator.
+ */
+public class SizedAggregator
```
Naming: given that this isn't actually an `Aggregator` object, it shouldn't end with `Aggregator`.  How about `AggregatorAndSize` instead?

kfaraz replied:
Sounds much better 👍

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/DimensionDictionary.java:
```diff
@@ -96,6 +99,12 @@ public int size()
 
+  public long sizeInBytes()
+  {
+    // TODO: size of map/list object itself?
```
Make sure to do something to remove the TODOs before merge.  Either implement or add a comment that explains what's missing and why it's potentially an issue or just delete.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/DimensionDictionary.java:
```diff
@@ -114,6 +123,11 @@ public int add(@Nullable T originalValue)
+
+      // Add size of the String and two of its references to the total size
+      // TODO: map entry overhead?
+      sizeInBytes.addAndGet(getObjectSize(originalValue) + Long.BYTES + Long.BYTES);
```
Instead of relying on the comment to be able to read what this code is doing, you can write it as

```
long sizeOfReference = Long.BYTES;
long sizeOfString = getObjectSize(originalValue);
sizeInBytes.addAndGet(sizeOfString + (2 * sizeOfReference));
```

Now, the actual code is saying the same thing that your comment was.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -225,13 +236,19 @@ protected AddToFactsResult addToFacts(
-    return ROUGH_OVERHEAD_PER_MAP_ENTRY + key.estimateBytesInMemory() + maxBytesPerRowForAggregators;
+    return ROUGH_OVERHEAD_PER_MAP_ENTRY
+           + key.estimateBytesInMemory()
+           + (useMaxMemoryEstimates ? maxBytesPerRowForAggregators : actualRowSizeForAggregators);
```
In reading this code I found myself wondering why push down the choice of which value to use so low and add an extra parameter here instead of change the call site of `estimateRowSizeInBytes` to check and pass the "correct" value?

kfaraz replied:
Fixed.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -240,22 +257,46 @@ public int getLastRowIndex()
     }
     rowContainer.set(null);
+
+    return useMaxMemoryEstimates ? 0 : totalInitialSizeBytes;
```
You've got the same check in multiple places.  I don't think you need the one here as `totalInitialSizeBytes` will be 0 in the `useMaxMemoryEstimates` case anyway.  But if you really want to ensure that the two code paths don't mix, move the `if(useMaxMemoryEstimates)` check to the top level and write the loop twice.  Once in a block that always returns `0` and another in a block that returns an aggregated sum.

kfaraz replied:
Fixed.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:
```diff
@@ -265,11 +306,12 @@ private void doAggregate(
       synchronized (agg) {
         try {
-          agg.aggregate();
+          totalIncrementalBytes += agg.aggregateWithSize();
```
This is going to call the `WithSize` version even when it's not supposed to (when it's using the max size).  Please adjust to only call the `WithSize` version when the flag is set.

kfaraz replied:
Fixed.

imply-cheddar commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorConfig.java:
```diff
@@ -34,6 +34,13 @@
+  /**
+   * Whether maximum memory usage should be considered in estimation.
+   */
+  default boolean isUseMaxMemoryEstimates() {
```
Should this really be a default method instead of becoming a real thing on the config object?

kfaraz commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregator.java:
```diff
@@ -21,23 +21,45 @@
+      sketchField.setAccessible(true);
+    }
+    catch (NoSuchFieldException | ClassNotFoundException e) {
+      LOG.error(e, "Could not initialize 'sketchField'");
```
Okay. For now, we can just fail loudly. The additional config can be done as a follow up.

kfaraz commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregator.java:
```diff
@@ -21,23 +21,45 @@
+  private Sketch sketch;
+
+  @Nullable
+  private static Field sketchField;
```
Updated.

imply-cheddar commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -478,6 +479,11 @@ public IncrementalIndexAddResult add(InputRow row) throws IndexSizeExceededExcep
 
+  public IncrementalIndexAddResult add(InputRow row, boolean skipMaxRowsInMemoryCheck) throws IndexSizeExceededException
+  {
+    return add(row, false, true);
```
This appears to be ignoring `skipMaxRowsInMemoryCheck` is that intentional?  Probably worth a comment as to why that's intentional if it is.

kfaraz replied:
Thanks a lot for catching this! I must have missed it.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/DimensionIndexer.java:
```diff
@@ -127,7 +127,7 @@
    */
   @Nullable
-  EncodedKeyComponentType processRowValsToUnsortedEncodedKeyComponent(@Nullable Object dimValues, boolean reportParseExceptions);
+  EncodedDimensionValue<EncodedKeyComponentType> processRowValsToUnsortedEncodedKeyComponent(@Nullable Object dimValues, boolean reportParseExceptions);
```
nit: javadoc params and return type needs updated.. but this isn't really new, it basically needed updated long ago, it is not always an array 😅

kfaraz replied:
Fixed.

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/plumber/RealtimePlumber.java:
```diff
@@ -271,6 +271,7 @@ private Sink getSink(long timestamp)
           config.getAppendableIndexSpec(),
           config.getMaxRowsInMemory(),
           config.getMaxBytesInMemoryOrDefault(),
+          true,
```
i've noticed a few places that aren't wired up to config are using true, so using the new behavior with no way out, is that intentional? this one in particular probably doesn't matter all that much these days (i hope), but i'm less sure about all of them, and it isn't consistent because i see some hard coded false in there too

kfaraz replied:
The value of the flag `useMaxMemoryEstimates = true` represents old behaviour.

The hard coding has been done only for the following classes:
- `RealtimePlumber` and related classes (hopefully not used anymore)
- `OnHeapIncrementalIndexBenchmark` (accidentally hardcoded to false, fixing this)

It should NOT be hardcoded to false in any of these cases.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java:
```diff
@@ -557,12 +573,17 @@ IncrementalIndexRowResult toIncrementalIndexRow(InputRow row)
         try {
-          dimsKey = indexer.processRowValsToUnsortedEncodedKeyComponent(row.getRaw(dimension), true);
+          final EncodedDimensionValue<?> encodedDimensionValue
+              = indexer.processRowValsToUnsortedEncodedKeyComponent(row.getRaw(dimension), true);
```
i guess due to the way this refactor was done (compared to aggs) there is no real way to turn off calculating the estimates, even if we aren't using them. maybe it doesn't matter if there is no/minimal performance impact

kfaraz replied:
Yes, I was not too sure about this either. I will take another and see if we can separate the two flows without too much duplication.

kfaraz replied:
Updated `DimensionDictionary` and `StringDimensionIndexer` to turn off estimations if not needed.

clintropolis reviewed (commented):
overall makes sense :+1:

@kfaraz have you done any measurement of the performance impact before/after this change so we know what we are getting ourselves into? Not sure which benchmarks would be most appropriate off the top of my head

kfaraz commented on processing/src/main/java/org/apache/druid/segment/DimensionDictionary.java:
```diff
@@ -160,4 +173,16 @@ public int getIdForNull()
     }
   }
+
+  private long getObjectSize(@Nonnull T object)
```
Thanks for pointing this out! I will see how we can make this cleaner.

kfaraz replied:
Used the `StringDimensionDictionary` suggestion, although I have not made it abstract so that implementations using the `DimensionDictionary` can continue to use it as the base concrete class.

kfaraz commented:
> have you done any measurement of the performance impact before/after this change so we know what we are getting ourselves into?

Thanks for the review, @clintropolis !
I don't have enough perf numbers yet which is why I have put the changes behind a flag for the time being.
I am working on the perf evaluation. Once we are satisfied with the numbers, we can get rid of the flag altogether.

abhishekagarwal87 commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java:
```diff
@@ -80,6 +81,14 @@ public Aggregator factorize(ColumnSelectorFactory metricFactory)
   }
 
+  @Override
+  public AggregatorAndSize factorizeWithSize(ColumnSelectorFactory metricFactory)
```
this needs some documentation.

kfaraz replied:
Added.

abhishekagarwal87 reviewed (commented):
Looks good overall. You should also document the approach that was finalized in the proposal and this implementation is based on. I think AggregatorFactory#factorizeWithSize is probably a good place for writing that doc and overall approach.

kfaraz commented on processing/src/main/java/org/apache/druid/query/aggregation/Aggregator.java:
```diff
@@ -42,6 +42,12 @@
 {
   void aggregate();
 
+  default long aggregateWithSize()
```
Thanks for the reminder, @abhishekagarwal87. I am adding javadocs wherever missing.

clintropolis replied:
what is the contract of this value, I don't see the word estimate in here, but think it probably should be... should implementors over-estimate if exact sizing is not possible or is under-estimating fine? Should there be a warning that the default estimate is used? (i imagine this would be very noisy if it is done per aggregate call... so don't really recommend doing it here or anything...)

kfaraz replied:
> Should there be a warning that the default estimate is used?

It would make sense to give this warning when `factorizeWithSize` is overridden but `aggregateWithSize` is not. In such a case, we might be significantly underestimating the memory usage.

As you said, doing it here might be noisy. A viable approach could be to have `factorizeWithSize` return a wrapper Aggregator which does not allow the regular `aggregate` (will be addressed in the subsequent PR).

kfaraz commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorAndSize.java:
```diff
@@ -0,0 +1,49 @@
+public class AggregatorAndSize
+{
+
+  // TODO: include default overhead for object sizes
```
Addressed in the caller.

kfaraz commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorAndSize.java:
```diff
@@ -0,0 +1,49 @@
+  // TODO: include default overhead for object sizes
+
+  private final Aggregator aggregator;
+  private final long initialSizeBytes;
```
Added. It should account for JVM object overhead too.

kfaraz commented on processing/src/main/java/org/apache/druid/segment/EncodedDimensionValue.java:
```diff
@@ -0,0 +1,48 @@
+{
+  @Nullable
+  private final K value;
+  private final long incrementalSizeBytes;
```
Added.

kfaraz commented:
> You should also document the approach that was finalized in the proposal and this implementation is based on. 

@abhishekagarwal87 , I have added an overview of the approach in `OnHeapIncrementalIndex`.
I have also added javadocs for the methods in `Aggregator` and `AggregatorFactory`, but they contain information only relevant to those methods.

Please let me know if this is sufficient.

clintropolis commented on processing/src/main/java/org/apache/druid/query/aggregation/AggregatorFactory.java:
```diff
@@ -69,6 +69,24 @@ public VectorAggregator factorizeVector(VectorColumnSelectorFactory selectorFact
+   *
+   * @return AggregatorAndSize which contains the actual aggregator and its initial size.
+   */
+  public AggregatorAndSize factorizeWithSize(ColumnSelectorFactory metricFactory)
```
same question about contract about returned sizes. Also, I wonder if there is anything we could do to make sure this method is overridden if `aggregateWithSize` is implemented, so that the initial size is not the max size...

kfaraz replied:
Updated the javadoc to advise on the required estimation.


> Also, I wonder if there is anything we could do to make sure this method is overridden if aggregateWithSize is implemented

I guess it is okay even if it isn't overridden because we would only be overestimating which would not cause failures, only somewhat poorer estimates.

kfaraz replied:
The other way around is probably more of an issue i.e. overriding `factorizeWithSize` but not `aggregateWithSize`. In such a case, `factorizeWithSize` would give a small initial size, which would never increase because `aggregateWithSize` would always return 0.

In either case, this issue is a problem if we use the new behaviour, i.e. `useMaxMemoryEstimates = false`. I guess we could address this is in a subsequent PR.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/StringDimensionIndexer.java:
```diff
@@ -60,20 +60,29 @@ private static String emptyToNullIfNeeded(@Nullable Object o)
   {
     final int[] encodedDimensionValues;
     final int oldDictSize = dimLookup.size();
+    final long oldDictSizeInBytes = useMaxMemoryEstimates ? 0 : dimLookup.sizeInBytes();
```
nit: this is only used inside of the else near the end of the method

kfaraz replied:
We need the size of the dictionary before adding the dimension values.
At the end, we take the final size of dictionary and check the diff.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/StringDimensionIndexer.java:
```diff
@@ -60,20 +60,29 @@ private static String emptyToNullIfNeeded(@Nullable Object o)
+      boolean useMaxMemoryEstimates
+  )
   {
+    super(useMaxMemoryEstimates ? new DimensionDictionary<>() : new StringDimensionDictionary());
```
nit: this seems strange/confusing, why wouldn't `StringDimensionIndexer` always use `StringDimensionDictionary` here? it seems like could just pass in a value of `false` to control the value of `computeOnHeapSize` instead of sometimes not using `StringDimensionDictionary`

kfaraz replied:
You are right, it is weird.
Fixed it.

clintropolis commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregator.java:
```diff
@@ -21,23 +21,43 @@
+  @Nullable
+  private static final Field SKETCH_FIELD;
+
+  static {
+    try {
+      SKETCH_FIELD = Class.forName("org.apache.datasketches.theta.UnionImpl")
+                          .getDeclaredField("gadget_");
+      SKETCH_FIELD.setAccessible(true);
+    }
+    catch (NoSuchFieldException | ClassNotFoundException e) {
+      throw new ISE(e, "Could not initialize SketchAggregator");
+    }
```
this seems worth a comment, and maybe a link to the code?

abhishekagarwal87 replied:
exception in static initialization blocks can surface as very weird errors. 
http://javaeesupportpatterns.blogspot.com/2012/07/javalangnoclassdeffounderror-how-to.html.

Maybe we can just move this initialization to the constructor? We need not worry too much about thread safety since it's ok even if SKETCH_FIELD gets constructed twice.

clintropolis reviewed (approved):
:+1:

abhishekagarwal87 commented:
@kfaraz CI failures might be legit. can you fix those before merging?
